app/server/appsmith-server/src/main/java/com/appsmith/server/services/UserOrganizationServiceImpl.java

```diff
@@ -254,6 +262,83 @@ public UserOrganizationServiceImpl(SessionUserService sessionUserService,
+            if (role.getUsername().equals(userRole.getUsername())) {
+                // User found in the organization.
+
+                if (role.getRole().equals(userRole.getRole())) {
+                    // No change in the role. Do nothing.
+                    return Mono.just(userRole);
+                } else if(role.getRole().equals(AppsmithRole.ORGANIZATION_ADMIN)) {
+                    // user is currently admin, check if this user is the only admin
+                    if(orgAdminCount == 1) {
+                        return Mono.error(new AppsmithException(AppsmithError.REMOVE_LAST_ORG_ADMIN_ERROR));
+                    }
+                }
+
+                // Step 1. Remove the existing role of the user from the organization
+                Mono<Organization> userRemovedOrganizationMono = this.removeUserRoleFromOrganizationGivenUserObject(organization, user);
+
+                // Step 2. Add the new role (if present) to the organization for the user
+                Mono<Organization> finalUpdatedOrganizationMono = userRemovedOrganizationMono;
+                if (userRole.getRoleName() != null) {
+                    // If a userRole name has been specified, then it means that the user's role has been modified.
+                    Mono<Organization> userAddedToOrganizationMono = userRemovedOrganizationMono
+                            .flatMap(organization1 -> this.addUserToOrganizationGivenUserObject(organization1, user, userRole));
+                    finalUpdatedOrganizationMono = userAddedToOrganizationMono.flatMap(addedOrganization -> {
+
+                        Map<String, String> params = new HashMap<>();
+                        params.put("Inviter_First_Name", currentUser.getName());
+                        params.put("inviter_org_name", organization.getName());
+                        params.put("inviteUrl", originHeader);
+                        params.put("user_role_name", userRole.getRoleName());
+
+                        Mono<Boolean> emailMono = emailSender.sendMail(user.getEmail(),
+                                "Appsmith: Your Role has been changed",
+                                UPDATE_ROLE_EXISTING_USER_TEMPLATE, params);
+                        return emailMono
+                                .thenReturn(addedOrganization);
+                    });
+                } else {
+                    // If the roleName was not present, then it implies that the user is being removed from the org.
+                    // Since at this point we have already removed the user from the organization,
+                    // remove the organization from recent org list of UserData
+                    // we also need to remove the org id from User.orgIdList
+                    finalUpdatedOrganizationMono = userRemovedOrganizationMono.flatMap(userDataRepository
+                            .removeOrgFromRecentlyUsedList(user.getId(), organization.getId())::thenReturn)
```

Comment:
Minor refactoring to read this better.

```suggestion
                    finalUpdatedOrganizationMono = userDataRepository
                            .removeOrgFromRecentlyUsedList(user.getId(), organization.getId())
                            .then(userRemovedOrganizationMono)
```